downstream.go

```diff
@@ -268,7 +268,7 @@ func (dc *downstreamConn) SendMessage(msg *irc.Message) {
 
 // marshalMessage re-formats a message coming from an upstream connection so
 // that it's suitable for being sent on this downstream connection. Only
-// messages that may appear in logs are supported.
```

Comment:
Hmm. So previously we had all the marshalling happening in one place, and now it's scattered around again. Can we move the marshalling logic in the upstream MODE handler here?

Oh, but then we need to parse the mode twice, which is annoying. Maybe we could have a function that turns a mode string into a `map[byte]bool` indicating the mode changes? Then we could use it from both places.

Reply:
To marshal the MODE message we need to know which mode letters are channel modes and which ones are membership prefixes. Then we marshal the mode params for a membership prefix but not for a channel mode.

The set of membership prefixes is stored in `upstreamConn.availableMemberships`, which is only available from an upstream, ie when we're connected (which makes sense: that's an `RPL_ISUPPORT` we parse from the upstream after registration and may vary between connections). That's incompatible with `downstreamConn.marshalMessage`, which expects a `network`, which makes sense because `marshalMessage` is also used to marshal messages from a network we're not currently connected to. In our case we'll only send the message when we're currently connected to that upstream.

Hence the need for specific logic here.

Reply:
All right, fair enough. I wonder if we can come up with a better design overall.